components/style/values/specified/position.rs

```diff
@@ -60,21 +125,80 @@ impl Position {
+            _ => (first_keyword, second_keyword),
         };
+
+        // Unwrap keywords from PositionComponent and wrap with Option.
+        let (wrapped_horiz_key, wrapped_vert_key) = if let Some(PositionComponent::Keyword(horiz_key)) = horiz_keyword {
+            if let Some(PositionComponent::Keyword(vert_key)) = vert_keyword {
+                (Some(horiz_key), Some(vert_key))
+            } else {
+                (Some(horiz_key), None)
+            }
+        } else {
+            (None, None)
+        };
+
         Ok(Position {
-            horizontal: horiz.to_length_or_percentage(),
-            vertical: vert.to_length_or_percentage(),
+            horiz_keyword: wrapped_horiz_key,
+            horiz_position: Some(horiz.unwrap().to_length_or_percentage()),
+            vert_keyword: wrapped_vert_key,
+            vert_position: Some(vert.unwrap().to_length_or_percentage()),
         })
     }
 
     pub fn parse(input: &mut Parser) -> Result<Position, ()> {
         let first = try!(PositionComponent::parse(input));
         let second = input.try(PositionComponent::parse)
-            .unwrap_or(PositionComponent::Center);
-        Position::new(first, second)
+            .unwrap_or(PositionComponent::Keyword(Keyword::Center));
+
+        // Try to parse third and fourth values
+        if let Ok(third) = input.try(PositionComponent::parse) {
+            if let Ok(fourth) = input.try(PositionComponent::parse) {
+                // Handle 4 value background position
+                Position::new(Some(second), Some(fourth), Some(first), Some(third))
+            } else {
+                // Handle 3 value background position
+                if let PositionCategory::LengthOrPercentage = category(first) {
+                    // "20px bottom 20%"
+                    Position::new(Some(first), Some(third), None, Some(second))
+                } else {
+                    if let PositionCategory::LengthOrPercentage = category(second) {
+                        if let PositionCategory::HorizontalKeyword = category(third) {
+                            // "bottom 10% right"
+                            Position::new(Some(second), None, Some(first), Some(third))
+                        } else {
+                            // "right 10px 50%"
+                            Position::new(Some(second), Some(third), Some(first), None)
+                        }
+                    } else {
+                        // "right bottom 10px"
+                        Position::new(None, Some(third), Some(first), Some(second))
+                    }
+                }
+            }
+        } else {
+            // Handle 2 value background position
+            Position::new(Some(first), Some(second), None, None)
+        }
+    }
+}
+
+impl ToCss for Keyword {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        match *self {
+            Keyword::Right => try!(dest.write_str("right ")),
```

Comment:
I didn't write them because of comment above. Just explicit keywords are storing.